File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        Map<Character, Integer> map = new HashMap<Character, Integer>();
        int maxLength = 0;

        int left=0;
        for(int right=0; right<s.length(); right++){
            while(map.containsKey(s.charAt(right)) && left <= left){
                map.remove(s.charAt(left));
                left++;
            }

            maxLength = Math.max(maxLength, right-left+1);
            map.put(s.charAt(right), right);
        }

        return maxLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        max_length = 0
        left = 0
        chars = {}

        for right in range(len(s)):

            while s[right] in chars and left <= chars[s[right]]:
                left += 1 
             
            max_length = max(max_length, right-left+1)
            chars[s[right]] = right

        return max_length